docs/design.md

```diff
@@ -1,92 +1,122 @@
-+ Managing transactions. e.g., to ensure atomicity of a transaction.
-+ Sanitizing input values recevied from the UI component.
-+ Providing a mechanism for checking access control rights.
+1. Request forwarded to the `ControllerServlet`.
+1. `ControllerServlet` uses the `ActionFactory` to generate the matching `Action` object, e.g. `InstructorHomePageAction`.
+1. `ControllerServlet` executes the action.
+1. The `Action` object checks the access rights of the user. If the action is allowed, it will be performed, interacting with the `Logic` component as necessary.
+1. The `Action` gathers the data required for the page into a `PageData` object, e.g. `InstructorHomePageData`, creates a `ShowPageResult` object by enclosing the `PageData` object created previously, and returns it to the `ControllerServlet`.
+1. `ControllerServlet` sends the result. In case of a `ShowPageResult`, this is equivalent to forwarding to the matching JSP page.
+1. The JSP page uses the data in the given `PageData` object to generate the HTML page.
+1. The response will then be sent back to the browser, which will render the page.
+
+The other 30% types of request may differ in the following manners:
+- At times, the user will request for data asynchronously via AJAX. In such cases, the actions return an `AjaxResult`. `PageData` is also created for this request and the data obtained is pre-processed for display in the browser; that is, the result of the AJAX request is not the raw data.
+- After performing certain actions, the browser may load another page (redirects to another page), possibly with the status of the previous action. In such cases, the action returns a `RedirectResult`, which simply instructs the browser to send a fresh request for the specified page. In such cases, no `PageData` object will be generated for the original request.
+- The result of some actions is downloading of a file (e.g. a feedback session report) or an image (e.g. profile picture). In such cases, the result type will be `FileDownloadResult` or `ImageResult`, and no `PageData` object will be generated.
+- In some cases, the request is not forwarded to `ControllerServlet`, but to some other `*Servlet` as specified in `web.xml`. In such cases, the request will be processed as specified in those `*Servlet` objects; no common routine is observed in these other `*Servlet` objects.
+- Some requests may not forwarded to any `*Servlet` at all, but rather to a specified resource file of the web application (e.g. JSP files).
+
+### Automated requests
+
+Automated requests are all requests sent automatically by the GAE server during specific periods of time.
+This type of request will be processed as follows:
+
+1. The source of the request will be checked for administrator privilege. If such privilege is absent (e.g. non-administrator users trying to invoke the automated actions), the request will be dropped and returned with a `403 Forbidden` status.
```

Comment:
"such a privilege" or "this privilege"
"...request will be dropped and a `403 Forbidden` status will be returned. "